File: c-sharp/Blackjack/src/Blackjack/InstanceFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Blackjack
{
    public class InstanceFactory
    {
    }


    public class InstanceFactory<T> : IDisposable
    {
        private Func<T> builder;
        private Func<T> overrideBuilder;

        public InstanceFactory(Func<T> builder)
        {
            this.builder = builder;
        }

        public T make()
        {
            if (overrideBuilder == null)
                return builder.Invoke();
            return overrideBuilder.Invoke();
        }

        public void Dispose()
        {
            overrideBuilder = null;
        }

        public InstanceFactory<T> overide(Func<T> builder) {
           this.overrideBuilder = builder;
           return this;
       }   
  }
}

File: java/src/main/java/navis/injection/InstanceFactory.java
package navis.injection;

public class InstanceFactory<T> implements Resettable {
    private final BuildFunction<T> builder;
    private BuildFunction<T> overrideBuilder;

    public InstanceFactory(BuildFunction<T> builder) {
        this.builder = builder;
    }

    public T make() {
        if (overrideBuilder == null)
            return builder.build();
        return overrideBuilder.build();
    }

    public Resettable override(BuildFunction<T> builder) {
        this.overrideBuilder = builder;
        return this;
    }

    @Override
    public void close() {
        overrideBuilder = null;
    }
}

